src/gil.rs

```diff
@@ -287,50 +287,60 @@ impl Drop for GILGuard {
     }
 }
 
+// Vector of PyObject(
+type PyObjVec = Vec<NonNull<ffi::PyObject>>;
+
 /// Thread-safe storage for objects which were inc_ref / dec_ref while the GIL was not held.
 struct ReferencePool {
-    pointers_to_incref: Mutex<Vec<NonNull<ffi::PyObject>>>,
-    pointers_to_decref: Mutex<Vec<NonNull<ffi::PyObject>>>,
+    dirty: atomic::AtomicBool,
+    // .0 is INCREFs, .1 is DECREFs
+    pointer_ops: Mutex<(PyObjVec, PyObjVec)>,
 }
 
 impl ReferencePool {
     const fn new() -> Self {
         Self {
-            pointers_to_incref: const_mutex(Vec::new()),
-            pointers_to_decref: const_mutex(Vec::new()),
+            dirty: atomic::AtomicBool::new(false),
+            pointer_ops: const_mutex((Vec::new(), Vec::new())),
         }
     }
 
     fn register_incref(&self, obj: NonNull<ffi::PyObject>) {
-        self.pointers_to_incref.lock().push(obj)
+        self.pointer_ops.lock().0.push(obj);
+        self.dirty.store(true, atomic::Ordering::Release);
     }
 
     fn register_decref(&self, obj: NonNull<ffi::PyObject>) {
-        self.pointers_to_decref.lock().push(obj)
+        self.pointer_ops.lock().1.push(obj);
+        self.dirty.store(true, atomic::Ordering::Release);
     }
 
     fn update_counts(&self, _py: Python) {
-        macro_rules! swap_vec_with_lock {
+        macro_rules! swap_vec {
             // Get vec from one of ReferencePool's mutexes via lock, swap vec if needed, unlock.
             ($cell:expr) => {{
-                let mut locked = $cell.lock();
                 let mut out = Vec::new();
-                if !locked.is_empty() {
-                    std::mem::swap(&mut out, &mut *locked);
+                if !$cell.is_empty() {
+                    std::mem::swap(&mut out, &mut $cell);
                 }
-                drop(locked);
                 out
             }};
         }
 
+        let prev = self.dirty.swap(false, atomic::Ordering::Acquire);
+        if !prev {
+            return;
+        }
+
+        let mut ops = self.pointer_ops.lock();
         // Always increase reference counts first - as otherwise objects which have a
         // nonzero total reference count might be incorrectly dropped by Python during
         // this update.
-        for ptr in swap_vec_with_lock!(self.pointers_to_incref) {
+        for ptr in swap_vec!(ops.0) {
             unsafe { ffi::Py_INCREF(ptr.as_ptr()) };
         }
 
-        for ptr in swap_vec_with_lock!(self.pointers_to_decref) {
+        for ptr in swap_vec!(ops.1) {
```

Comment:
We need to be careful here. Decrementing references can run arbitrary Python drop code, which can lead to GIL release which may eventually deadlock because we hold the lock here. (I think this mutex is also non reentrant so may deadlock on a single thread.)

After swapping `ops.1` we should release the lock before decreasing any references.

Reply:
Good call, should be good now.